portal-bridge/src/types/state/trie_walker.rs

```diff
@@ -0,0 +1,190 @@
+                                    .clone(),
+                                    Some(node_key),
+                                    extension.prefix.get_data().to_vec(),
+                                ),
+                            );
+                        }
+                    }
+                }
+                Node::Branch(branch) => {
+                    let branch = branch.read().expect("Reading a branch should work");
+                    for (i, child) in branch.children.iter().enumerate() {
+                        // We look for hash nodes in order to connect them to the root. If this node
+                        // is not a hash node, then neither is any of its children.
+                        // We know this because any node that has a hash node as it's descendant
+                        // would also become hash node, because its encoding would be longer than 32
+                        // bytes.
+                        if let Node::Hash(hash_node) = child {
+                            //Only process provided nodes (they belong to the partial trie that we
+                            // care about)
+                            if nodes.contains_key(&hash_node.hash) {
+                                stack.push(hash_node.hash);
+                                trie_walker_nodes.insert(
+                                    hash_node.hash,
+                                    TrieWalkerNode::new(
+                                        nodes
+                                    .get(&hash_node.hash)
+                                    .expect("The stack should only contain nodes that are in the changed nodes")
+                                    .clone(),
+                                        Some(node_key),
+                                        vec![i as u8],
+                                    ),
+                                );
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // Ensure we discovered all nodes (root is not present in the trie_walker_nodes)
+        ensure!(nodes.len() == trie_walker_nodes.len(), "We expect the node length and trie_walker_nodes length to be the same, as we should have discovered all nodes in the trie. If this fails it implies an invalid trie was provided with an unnavigable path.");
+
+        Ok(trie_walker_nodes)
+    }
+
+    pub fn get_proof(&self, node_hash: B256) -> AccountProof {
+        // Build path and proof from the node to the root and reverse it at the end.
+        let mut reverse_path = vec![];
+        let mut reverse_proof = vec![];
+        let mut next_node: Option<B256> = Some(node_hash);
+        while let Some(current_node) = next_node {
+            let Some(parent) = self.nodes.get(&current_node) else {
+                break;
```

Comment:
nit: I believe that this should be consider an error now (because even root is stored in the `self.nodes`. I would `bail!`/`panic!` here, but if you think it's too much, we should at least print error.